### PR TITLE
Conditionally promisified API calls for alternative syntax

### PR DESCRIPTION
**Pull request by Eluvade**
Hi!
I've been working with your library for a while, as I was hooking it up to my high-frequency autonomous trading bot. Due to the complexity of my project, I prefer using the newer `async` / `await` syntax to increase maintainability by avoiding nested callbacks. It would have probably been better to re-write the library by using axios, but I found a solution that should not break functionality for any dependants of the library nor impact performance. Admittedly, the code size has increased, but I don't think I could have further optimized much of the added code. 
Keeping the old functionality intact, merging should not break functionality for anyone. The feature I've added is the ability to use `.then()` / `.catch()` and `async` / `await` syntax when calling any of the functions that make an API request. In order to minimize meddling with the current code structure, I've decided to return a promise whenever a callback isn't passed to a function that interacts with the API endpoint.

Please take a look at an [example](https://github.com/eluvade/node-binance-api/blob/master/examples/advanced.md#asynchronous-syntax-options) of calling the same `node-binance-api` function with three different syntaxes.

However, my work isn't entirely done. I have not touched the functions belonging to the websockets object. Before I commit any more work, I'd like to get some input from you to see whether this feature is something you'd like in the first place. I'm aware you're no longer maintaining the repository regularly, so I'll try not troubling you any further. But in case you'd like to reach me, here's my e-mail:  bunny.eluvade@gmail.com

Have a great day!

TL;DR Promisified functions that make API requests while still allowing to use regular callbacks instead.

### CODE REVIEW

**Comment by Eluvade** [comment]
I'm pretty sure that test.js is rather outdated. I haven't worked with mocha since I was last testing a solidity smart contract, but I might get to it next week.

**Comment by Eluvade** [comment]
There's one more thing I've been trying to figure out a solution for; enabling `useServerTime` makes an asynchronous request but the function already returns the instantiated Binance object. Due to this, `timeOffset` in `Binance.info` has not yet been set at the time of initialization so the current solution was to wrap the entry point of one's application flow within the `.options()` callback parameter. I've tried returning a promise which is resolved with the Binance object once the offset is set. However, I wasn't able to call any of the Binance object's methods after receiving it from the promise. I'll take a closer look tomorrow, I guess.

**Comment by jaggedsoft** [comment]
I really like what you're doing, you have my full support as these features have been requested for a long time.

**Comment by jaggedsoft** [comment]
> There's one more thing I've been trying to figure out a solution for; enabling `useServerTime` makes an asynchronous request but the function already returns the instantiated Binance object. Due to this, `timeOffset` in `Binance.info` has not yet been set at the time of initialization so the current solution was to wrap the entry point of one's application flow within the `.options()` callback parameter. I've tried returning a promise which is resolved with the Binance object once the offset is set. However, I wasn't able to call any of the Binance object's methods after receiving it from the promise. I'll take a closer look tomorrow, I guess.

My solution for useServerTime in the past has been to do it like this

```js
const Binance = require('node-binance-api');
const binance = new Binance().options({
  APIKEY: '',
  APISECRET: ''
});

// Callback once useServerTime is synchronized
binance.useServerTime(function() {
	// Entry point of your application. You can call whatever function you want and it will work
});
```

**Comment by jaggedsoft** [comment]
I'm ready to merge all your changes when you give me the word. If it's stable now, I'll create a release 👍

**Comment by Eluvade** [comment]
> > There's one more thing I've been trying to figure out a solution for; enabling `useServerTime` makes an asynchronous request but the function already returns the instantiated Binance object. Due to this, `timeOffset` in `Binance.info` has not yet been set at the time of initialization so the current solution was to wrap the entry point of one's application flow within the `.options()` callback parameter. I've tried returning a promise which is resolved with the Binance object once the offset is set. However, I wasn't able to call any of the Binance object's methods after receiving it from the promise. I'll take a closer look tomorrow, I guess.
> 
> My solution for useServerTime in the past has been to do it like this
> 
> ```js
> const Binance = require('node-binance-api');
> const binance = new Binance().options({
>   APIKEY: '',
>   APISECRET: ''
> });
> 
> // Callback once useServerTime is synchronized
> binance.useServerTime(function() {
> 	// Entry point of your application. You can call whatever function you want and it will work
> });
> ```

Ah, I figured as much. I'll leave it at that then. Don't merge yet though. I'll try finishing up the websocket functions by tomorrow.

**Comment by Eluvade** [comment]
I was looking for potential ways to promisify websocket connection returns. This library could come in handy, but it would be a lot of work integrating it into the project: https://github.com/vitalets/websocket-as-promised 
However, as a concept, promises don't make much sense with a websocket connection. We should either leave it as plain callbacks or promisify only subscription to a socket by resolving the promise within this line 
```js
ws.on('open', handleSocketOpen.bind(ws, opened_callback));
```

**Comment by jaggedsoft** [comment]
I agree, websockets-as-promised is really cool but since websockets are callback based they're probably more convenient left that way. Thank you for all of your help, really nice work.

**Comment by yosietserga** [comment]
Hi!

This changes will break the backward compatibility? I agree with all in this conversation, but I already made my own app proxy to use promises upon websockets and I am using this library as binance api platform. 

That I want trying to say is that the issue/fix in this converstion can be done by each developer in their own projects.

I use the next flow:

Connect to WS -> Wrap WS connection passing callback function -> callback function fill data registry and return promise -> Outside, I can use the app registry with data updated or get promise return data and do anything that I need

In context, I called it workers and each worker has a specific function. One worker connect to websocket and fill app registry  and another worker process that information.

- Doorman Worker: this connect to WS and fill/update data
- Buyer Worker: using the data filled by doorman, determine when to buy 
- Trader Worker: using the data filled by doorman and buyer, determine when to sell or until when to hold

**Comment by Eluvade** [comment]
> Hi!
> 
> This changes will break the backward compatibility? I agree with all in this conversation, but I already made my own app proxy to use promises upon websockets and I am using this library as binance api platform.
> 
> That I want trying to say is that the issue/fix in this converstion can be done by each developer in their own projects.
> 
> I use the next flow:
> 
> Connect to WS -> Wrap WS connection passing callback function -> callback function fill data registry and return promise -> Outside, I can use the app registry with data updated or get promise return data and do anything that I need
> 
> In context, I called it workers and each worker has a specific function. One worker connect to websocket and fill app registry and another worker process that information.
> 
> * Doorman Worker: this connect to WS and fill/update data
> * Buyer Worker: using the data filled by doorman, determine when to buy
> * Trader Worker: using the data filled by doorman and buyer, determine when to sell or until when to hold

Hey! Thank you for sharing your insight. 

The changes I'm proposing **should not** break the application flow for any of the module's existing dependants. Please do take a look in case I've overlooked something.

I believe a lot of developers using this library have come up with their own solutions and implementations to deal with the syntax limitations we've been discussing. I too, have been using my own solution to promisifying function returns within my application. Considering this has to be done as an entire level above the the library, it decreases performance and maintainability while increasing development time. 

As for your inquiry about the functions belonging to the `websockets` object; they haven't been changed. At least for now.

**Comment by Eluvade** [comment]
> I agree, websockets-as-promised is really cool but since websockets are callback based they're probably more convenient left that way. Thank you for all of your help, really nice work.

I have concluded my work on this repository for now. You may merge at your own convenience.

**Comment by jaggedsoft** [comment]
This week I don't have time to review the code, but I can push a release if you think it's good to go

**Comment by Eluvade** [comment]
> This week I don't have time to review the code, but I can push a release if you think it's good to go

I think it is, but I'd prefer you to review the code just in case; so take your time.

**Comment by waterdrop01** [comment]
Hello, I'd love to get this merged :)

**Comment by jaggedsoft** [comment]
Life has been extremely busy for me, I won't have time to review this so if you say it's good I'll just publish it.

**Comment by Eluvade** [comment]
> Life has been extremely busy for me, I won't have time to review this so if you say it's good I'll just publish it.

Okay, please publish it then.

**Comment by jaggedsoft** [comment]
Thanks for your contribution!

**Comment by jaggedsoft** [comment]
Thanks for your contribution!
